### PR TITLE
Translation Updates

### PR DESCRIPTION
**Pull request by james-prado**
Closes https://github.com/MyCryptoHQ/MyCrypto/issues/1285

#### Description

Move raw text to translation json files. Update translation functions. Remove old translation strings.

#### Changes

- Update `translate()` and `translateRaw()` functions
	- nows accept variables in translation strings
	- variables can be placed into any translation string, and update before being rendered as markdown or raw text
- Move all* strings of text to `translation/*.json` files
	- Some names and technical terms have purposely not been translated. MyCrypto, Ledger, Infura, etc... are specific names and don't need to be translated. ABI, bytes32, uint256, etc... are technical terms that will be **very** difficult to translate properly (maybe even impossible) without professional translation
- Replace all raw text with new translation functions
- Get selected language from localstorage instead of redux 
	- some strings are translated before the redux store has actually initialized. Trying to access the redux store before this will throw a vague error msg and stop everything from rendering. To avoid this, I get the selected language from localstorage, the same value that gets put into the rehydrated redux state. 

#### Testing 

Go to https://mycryptobuilds.com/39ebd4f226963999582759e3166751c7528b5065

I've added emoji's to the end of translated and non-translated strings to visually spot raw text strings. 

🎉 = translated 
☠ = not translated

This method does produces false negatives when putting a variable or link in a string to be translated, but it's the best method I've got for finding untranslated strings. 

Strings with ☠ at the end, or no emoji at all definitely aren't translated. Just let me know if you find one and where the string is, I'll add it into the translation json files.

### CODE REVIEW

File: common/components/Footer/index.tsx [hunk 1]
```diff
@@ -127,6 +129,23 @@ export default class Footer extends React.PureComponent<Props, State> {
     isDisclaimerOpen: false
   };
 
+  public componentDidMount() {
```
Comment (dternyak) [inline common/components/Footer/index.tsx]: Was this a debug tool?

Comment (james-prado) [reply common/components/Footer/index.tsx]: yes

File: common/components/Footer/index.tsx [hunk 2]
```diff
@@ -165,30 +184,27 @@ export default class Footer extends React.PureComponent<Props, State> {
...
-                <a onClick={this.toggleModal}>Disclaimer</a>
+                <a onClick={this.toggleModal}>{translateRaw('DISCLAIMER')}</a>
               </div>
               <div className="Footer-about-legal-text">v{VERSION}</div>
             </div>
           </div>
 
           <div className="Footer-support Footer-section">
-            <h5 className="Footer-support-title">Support Us & Our Friends</h5>
+            <h5 className="Footer-support-title">{translateRaw('FOOTER_AFFILIAT_TITLE')}</h5>
```
Comment (wbobeirne) [inline common/components/Footer/index.tsx]: *AFFILIATE

File: common/components/Footer/index.tsx [hunk 3]
```diff
@@ -198,12 +214,16 @@ export default class Footer extends React.PureComponent<Props, State> {
             </div>
 
             <div className="Footer-support-donate">
-              <div className="Footer-support-donate-currency">Donate ETH</div>
+              <div className="Footer-support-donate-currency">
+                {translateRaw('DONATE_CURRENCY') + ' ETH'}
```
Comment (wbobeirne) [inline common/components/Footer/index.tsx]: This is a case for variables, some languages may put the subject (ETH) before the verb (Donate)

File: common/components/GenerateKeystoreModal/index.tsx [hunk 4]
```diff
@@ -91,7 +91,7 @@ export default class GenerateKeystoreModal extends React.Component<Props, State>
               className="GenKeystore-button btn btn-primary btn-block"
               disabled={!isPrivateKeyValid || !isPasswordValid}
             >
-              {translate('Generate Keystore File')}
+              {translateRaw('Generate Keystore File')}
```
Comment (wbobeirne) [inline common/components/GenerateKeystoreModal/index.tsx]: These should be translation keys, not raw text.

File: common/components/GenerateKeystoreModal/index.tsx [hunk 5]
```diff
@@ -108,7 +108,7 @@ export default class GenerateKeystoreModal extends React.Component<Props, State>
               aria-describedby={translateRaw('x_KeystoreDesc')}
               download={keystoreFile.filename}
             >
-              {translate('Download Keystore File')}
+              {translateRaw('Download Keystore File')}
```
Comment (wbobeirne) [inline common/components/GenerateKeystoreModal/index.tsx]: Should be translation keys, not raw text.

File: common/components/TXMetaDataPanel/TXMetaDataPanel.tsx [hunk 6]
```diff
@@ -114,7 +114,7 @@ class TXMetaDataPanel extends React.Component<Props, State> {
             <div className="help-block">
               <a className="Gas-toggle" onClick={this.toggleAdvanced}>
                 {showAdvanced
-                  ? `- ${translateRaw('Back to simple')}`
+                  ? `- ${translateRaw('ACTION_4')}`
                   : `+ ${translateRaw('Advanced Settings')}`}
```
Comment (wbobeirne) [inline common/components/TXMetaDataPanel/TXMetaDataPanel.tsx]: Should be a translation key and not raw text.

File: common/components/TXMetaDataPanel/components/AdvancedGas.tsx [hunk 7]
```diff
@@ -74,7 +74,7 @@ class AdvancedGas extends React.Component<Props, State> {
             <div className="input-group-wrapper AdvancedGas-gas-price">
               <label className="input-group">
                 <div className="input-group-header">
-                  {translate('OFFLINE_Step2_Label_3')} (gwei)
+                  {translateRaw('OFFLINE_Step2_Label_3')} (gwei)
```
Comment (wbobeirne) [inline common/components/TXMetaDataPanel/components/AdvancedGas.tsx]: I'd like to see a better key for this one, the `OFFLINE_stepX` keys are hard to parse at a glance.

File: common/components/TXMetaDataPanel/components/SimpleGas.tsx [hunk 8]
```diff
@@ -108,8 +108,8 @@ class SimpleGas extends React.Component<Props> {
               disabled={isGasEstimating}
             />
             <div className="SimpleGas-slider-labels">
-              <span>{translate('Cheap')}</span>
-              <span>{translate('Fast')}</span>
+              <span>{translateRaw('Cheap')}</span>
+              <span>{translateRaw('Fast')}</span>
```
Comment (wbobeirne) [inline common/components/TXMetaDataPanel/components/SimpleGas.tsx]: Should be keys and not raw strings.

File: common/components/WalletDecrypt/WalletDecrypt.tsx [hunk 9]
```diff
@@ -133,311 +135,319 @@ const SECURE_WALLETS = Object.values(SecureWalletName);
...
-      helpLink: 'https://doc.satoshilabs.com/trezor-apps/mew.html'
-    },
-    [InsecureWalletName.KEYSTORE_FILE]: {
-      lid: 'x_Keystore2',
-      example: 'UTC--2017-12-15T17-35-22.547Z--6be6e49e82425a5aa56396db03512f2cc10e95e8',
-      component: KeystoreDecrypt,
-      initialParams: {
-        file: '',
-        password: ''
+const WalletDecrypt = withRouter<Props>(
```
Comment (wbobeirne) [inline common/components/WalletDecrypt/WalletDecrypt.tsx]: You can spare yourself the indentation by moving the withRouter call to the export below, ala connect.

File: common/components/WalletDecrypt/components/InsecureWalletWarning.tsx [hunk 10]
```diff
@@ -61,38 +62,31 @@ export class InsecureWalletWarning extends React.Component<Props, State> {
...
+              var_run_local_article: knowledgeBaseURL + HELP_ARTICLE.RUNNING_LOCALLY
+            })}
           </li>
           <li>
-            Reading{' '}
-            <HelpLink article={HELP_ARTICLE.SECURING_YOUR_ETH}>
-              How to Protect Yourself and Your Funds
-            </HelpLink>
+            {translateMarkdown('INSECURE_WALLET_RECOMMEND_3', {
+              var_secure_your_eth_article: knowledgeBaseURL + HELP_ARTICLE.SECURING_YOUR_ETH
```
Comment (wbobeirne) [inline common/components/WalletDecrypt/components/InsecureWalletWarning.tsx]: I've noticed most of these translation variables are prefixed with `var_`, is there any reason for that?

File: common/components/WalletDecrypt/components/Mnemonic.tsx [hunk 11]
```diff
@@ -78,7 +78,7 @@ class MnemonicDecryptClass extends PureComponent<Props, State> {
               className="btn btn-primary btn-lg"
               disabled={!isValidMnemonic}
             >
-              {translate('Choose Address')}
+              {translateRaw('Choose Address')}
```
Comment (wbobeirne) [inline common/components/WalletDecrypt/components/Mnemonic.tsx]: Should be a key, not a raw string.

File: common/components/WalletDecrypt/components/Web3.tsx [hunk 12]
```diff
@@ -11,14 +11,14 @@ export const Web3Decrypt: React.SFC<Props> = ({ onUnlock }) => (
...
-        {translate('ADD_MetaMask')}
+        {translateRaw('ADD_MetaMask')}
       </button>
     </div>
 
     <div>
       <NewTabLink
         className="Web3Decrypt-install btn btn-sm btn-default btn-block"
-        content={translate('Download MetaMask')}
+        content={translateRaw('Download MetaMask')}
```
Comment (wbobeirne) [inline common/components/WalletDecrypt/components/Web3.tsx]: Should be a key, not a raw string. Might be a good idea to make this a variable'd translation so we could make it "Download $thing"

File: common/components/ui/UnlockHeader.tsx [hunk 13]
```diff
@@ -44,7 +44,7 @@ export class UnlockHeader extends React.PureComponent<Props, State> {
             >
               <span>
                 <span className="hidden-xs UnlockHeader-open-text">
-                  {translate('Change Wallet')}
+                  {translateRaw('Change Wallet')}
```
Comment (wbobeirne) [inline common/components/ui/UnlockHeader.tsx]: Should be a key, not raw text

File: common/config/data.tsx [hunk 14]
```diff
@@ -48,7 +48,7 @@ export const gasPriceDefaults: GasPriceSetting = {
 
 export const MINIMUM_PASSWORD_LENGTH = 12;
 
-export const knowledgeBaseURL = 'https://support.mycrypto.com';
+export const knowledgeBaseURL = 'https://support.mycrypto.com/';
```
Comment (wbobeirne) [inline common/config/data.tsx]: Double check this one, you're going to end up with a _lot_ of double slashes in a lot of places, which will be an invalid URL. I'd advise leaving it as-is.

File: common/containers/Tabs/Contracts/components/Deploy.tsx [hunk 15]
```diff
@@ -26,12 +26,12 @@ class DeployClass extends Component<DispatchProps> {
       <main className="Deploy Tab-content-pane" role="main">
         <button className="Deploy-field-reset btn btn-default btn-sm" onClick={this.changeWallet}>
           <i className="fa fa-refresh" />
-          {translate('Change Wallet')}
+          {translateRaw('Change Wallet')}
```
Comment (wbobeirne) [inline common/containers/Tabs/Contracts/components/Deploy.tsx]: Should be a key, not raw text

File: common/containers/Tabs/Contracts/index.tsx [hunk 16]
```diff
@@ -18,11 +18,11 @@ interface Props {
 const tabs = [
   {
     path: 'interact',
-    name: translate('Interact')
+    name: translateRaw('Interact')
   },
   {
     path: 'deploy',
-    name: translate('Deploy')
+    name: translateRaw('Deploy')
```
Comment (wbobeirne) [inline common/containers/Tabs/Contracts/index.tsx]: Should be keys, not raw text

File: common/containers/Tabs/GenerateWallet/components/Keystore/EnterPassword.tsx [hunk 17]
```diff
@@ -108,11 +110,11 @@ export default class EnterPassword extends Component<Props, State> {
       const { password, passwordValidation } = this.state;
 
       if (password.length < MINIMUM_PASSWORD_LENGTH) {
-        feedback = `Password must be ${MINIMUM_PASSWORD_LENGTH}+ characters`;
+        feedback = translateRaw('INPUT_PASSWORD_PLACEHOLDER');
       } else if (passwordValidation && passwordValidation.feedback) {
-        feedback = `This password is not strong enough. ${passwordValidation.feedback.warning}.`;
+        feedback = translateRaw('WEAK_PASSWORD') + ' ' + passwordValidation.feedback.warning;
```
Comment (wbobeirne) [inline common/containers/Tabs/GenerateWallet/components/Keystore/EnterPassword.tsx]: We should potentially figure out a way to localize these password feedback warnings.

File: common/containers/Tabs/SendTransaction/components/RequestPayment.tsx [hunk 18]
```diff
@@ -123,7 +123,7 @@ class RequestPayment extends React.Component<Props, {}> {
 
           {!!eip681String.length && (
             <div className="row form-group">
-              <label className="RequestPayment-title">{translate('Payment QR & Code')}</label>
+              <label className="RequestPayment-title">{translateRaw('Payment QR & Code')}</label>
```
Comment (wbobeirne) [inline common/containers/Tabs/SendTransaction/components/RequestPayment.tsx]: Should be a key, not raw text

File: common/containers/Tabs/SendTransaction/index.tsx [hunk 19]
```diff
@@ -39,24 +39,24 @@ class SendTransaction extends React.Component<Props> {
...
       {
         path: 'send',
-        name: translate('NAV_SendEther'),
+        name: translateRaw('NAV_SendEther'),
         disabled: !!wallet && !!wallet.isReadOnly
       },
       {
         path: 'request',
-        name: translate('Request Payment'),
+        name: translateRaw('Request Payment'),
```
Comment (wbobeirne) [inline common/containers/Tabs/SendTransaction/index.tsx]: Should be a key and not raw text

File: common/containers/Tabs/SendTransaction/index.tsx [hunk 20]
```diff
@@ -39,24 +39,24 @@ class SendTransaction extends React.Component<Props> {
...
-        name: translate('NAV_ViewWallet')
+        name: translateRaw('NAV_ViewWallet')
       }
     ];
 
     return (
       <TabSection>
         <section className="Tab-content">
-          <UnlockHeader title={translate('Account')} showGenerateLink={true} />
+          <UnlockHeader title={translateRaw('Account')} showGenerateLink={true} />
```
Comment (wbobeirne) [inline common/containers/Tabs/SendTransaction/index.tsx]: Should be a key and not raw text

File: common/containers/Tabs/SignAndVerifyMessage/components/SignMessage/index.tsx [hunk 21]
```diff
@@ -52,20 +51,20 @@ export class SignMessage extends Component<Props, State> {
               onClick={this.changeWallet}
             >
               <i className="fa fa-refresh" />
-              {translate('Change Wallet')}
+              {translateRaw('Change Wallet')}
```
Comment (wbobeirne) [inline common/containers/Tabs/SignAndVerifyMessage/components/SignMessage/index.tsx]: Should be a key and not raw text

File: common/containers/Tabs/Swap/components/LiteSend/Fields.tsx [hunk 22]
```diff
@@ -31,7 +31,7 @@ class FieldsClass extends Component<Props> {
               onClick={this.changeWallet}
             >
               <i className="fa fa-refresh" />
-              {translate('Change Wallet')}
+              {translateRaw('Change Wallet')}
```
Comment (wbobeirne) [inline common/containers/Tabs/Swap/components/LiteSend/Fields.tsx]: Should be a key and not raw text. Might also consider making the change wallet button a component if that makes any sense, I see it a whole lot.

File: common/utils/findandreplace.js [hunk 23]
```diff
@@ -0,0 +1,648 @@
...
+          lastNode.parentNode.replaceChild(matchEndNode, lastNode);
+        });
+
+        return lastNode;
+      }
+    }
+  };
+
+  return exposed;
+});
```
Comment (wbobeirne) [inline common/utils/findandreplace.js]: This looks a lot like a module. Should we be pulling in a dependency instead?

File: common/translations/lang/en.json [hunk 24]
```diff
@@ -193,10 +222,12 @@
     "BULK_Label_1": "Number of Wallets To Generate ",
     "BULK_Label_2": "Generate Wallets ",
     "BULK_SuccessMsg": "Success! Your wallets have been generated. ",
-    "SEND_addr": "To Address ",
+    "SEND_addr": "To ",
```
Comment (wbobeirne) [inline common/translations/lang/en.json]: We should add a `SEND_addr_short` instead, I think we may want to keep this key around.

File: common/components/WalletDecrypt/WalletDecrypt.tsx (hunk 9 above)
Comment (james-prado) [reply common/components/WalletDecrypt/WalletDecrypt.tsx]: I tried that but I wasn't able to get the types to work properly, not sure why :/ I'll have another look and see if I can clean it up

File: common/components/WalletDecrypt/components/InsecureWalletWarning.tsx (hunk 10 above)
Comment (james-prado) [reply common/components/WalletDecrypt/components/InsecureWalletWarning.tsx]: just to denote that that represents a variable. Using `var_` as a prefix is just a preference.

File: common/utils/findandreplace.js (hunk 23 above)
Comment (james-prado) [reply common/utils/findandreplace.js]: It's just used for debugging. It helps adds the emoji's to strings. It'll be removed with the rest of the debugging stuff.

**Comment by james-prado** [comment]
@wbobeirne fixed up most of the suggested changes.

I'd like to standardize the naming of translation keys but right now I can't think of any effective way to do it at scale.

I was able to make all the keys uppercase with some regex find and replace, and a little manual work, but we'll need to come up with a better method to update the rest of the strings

**Comment by james-prado** [comment]
Addressing @wbobeirne comment on having a conditional return type for `translate()`, ie. 
```
function translate(markdown: true): React.Element<any>;
function translate(markdown: false): string;
```

I agree, ideally we shouldn’t have to know the content of the translation string to translate it properly and provide the correct return type, but I can’t find any better way to solve this issue. Returning `React.Element<any>` all the time doesn’t work because other components / module require strings as arguments, and returning strings all the time doesn’t work because then you can’t return proper markdown. There really isn’t any good solution to this other than changing the return type if you know you’re using a markdown string.

**Comment by wbobeirne** [comment]
I think the solution for that issue was what we had before. `translate` is always a React element, `translateRaw` is always a string. If there's a component that was only accepting strings before, it should accept React elements as well if and only if it's rendering that prop in a React element friendly context. Otherwise, it's good that it only accepts a string because we should be using `translateRaw` in that instance.

```jsx
// If it goes in a <p>, it can _always_ be a react element.
<p>{translate('x_Description')}</p>

// Placeholders should _always_ be strings. If this were a prop,
// it should be typed to only accept strings.
<input placeholder={translateRaw('x_Placeholder')}/>
```

React elements for all translated text is the pattern that most libraries use, like [react-intl](https://github.com/yahoo/react-intl) and [react-i18n-next](https://github.com/i18next/react-i18next). They also typically provide an escape hatch string returning function for things like placeholder text.

I feel like if we're having difficulty coming to consensus, it might be time to use a third party library to solve our translation rather than trying to reinvent the wheel.

**Comment by dternyak** [comment]
I've taken some time to review both approaches, and I have a few thoughts. 

It took me some time to try to determine the objective benefits and drawbacks at each approach, but I believe I can summarize them as follows. I may have missed a few items, so please feel free to add as needed:

### Single `translate` function
Benefits:
1. No need for multiple imports of `translate` and `translateRaw`
2. [I'm sure I'm missing something here, but I couldn't determine additional benefits]

Drawbacks:
1. Call signatures have to include `, {}, true` to support markdown
2. (subjective) Slightly less clear overall API. `translate` always returning a component and `translateRaw` always returning a string gives you a clear reason for choosing each, whereas a parameter signature determination of return value is less clear.

### Existing approach.
Benefits:
1. Clearer API (call signature is simpler, return value is simpler as described above)
2. Supporting markdown if the origin string changes requires no changes unless that string is being used as placeholders / string required properties. I think the existing API could be modified to include `forceString` as an optional argument defaulting to false if this becomes a significant hurdle.

Drawbacks:
1. Requires multiple imports of `translate` and `translateRaw`
1. [Please insert what I'm missing]

Overall, with what information I've been able to gather, it seems like the changes are not overall beneficial and in some cases are regressions. I would prefer to stick with the current approach as well, but I think there are some points I am probably missing, so it would be worth having @wbobeirne and @james-prado help updates the above breakdown to ensure we have all available information.

**Comment by dternyak** [comment]
screen shot 2018-03-21 at 1 09 33 pm

Issue on the account view.
